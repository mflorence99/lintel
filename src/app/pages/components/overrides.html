<div
  [formGroup]="overridesForm"
  class="overrides">

  <lintel-input-array
    [enabled]="lintel.isEnabled"
    [formControl]="control"
    *ngFor="let control of overridesForm.controls.files['controls']; let ix = index">

    <div prefix>
      {{ ix + 1 }}
    </div>

    <div suffix>
      <fa-icon
        (click)="showContextMenu($event, ix)"
        [contextMenu]="contextMenu"
        [contextMenuSubject]="ix"
        [icon]="['fas', 'cog']">
      </fa-icon>
    </div>

  </lintel-input-array>

</div>

<!-- Tools menu -->

<context-menu>

  <ng-template
    contextMenuItem
    (execute)="execute($event.item, 'add')"
    [enabled]="lintel.isEnabled">
    <p>Add a new override</p>
  </ng-template>

  <ng-template
    contextMenuItem
    (execute)="execute($event.item, 'delete')"
    [enabled]="lintel.isEnabled">
    <p>Delete this override</p>
  </ng-template>

</context-menu>
