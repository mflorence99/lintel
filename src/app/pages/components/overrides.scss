:host {
  display: block;
}

.overrides {
  display: grid;
  grid-row-gap: 0.75rem;
  grid-template-columns: repeat(auto-fit, 1fr);

  [prefix] {
    margin-right: 0.5rem;
  }

  [suffix] {
    margin-left: 0;
    opacity: var(--lintel-opacity-icon);
  }
}
