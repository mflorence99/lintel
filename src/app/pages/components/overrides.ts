import { Actions } from '@ngxs/store';
import { ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { ConfigsState } from '../../state/configs';
import { ContextMenuComponent } from 'ngx-contextmenu';
import { ContextMenuService } from 'ngx-contextmenu';
import { DestroyService } from '../../services/destroy';
import { ExtensionsState } from '../../state/extensions';
import { FormArray } from '@angular/forms';
import { FormBuilder } from '@angular/forms';
import { FormControl } from '@angular/forms';
import { FormGroup } from '@angular/forms';
import { LintelState } from '../../state/lintel';
import { OnInit } from '@angular/core';
import { SchemaState } from '../../state/schema';
import { SelectionState } from '../../state/selection';
import { ViewChild } from '@angular/core';

import { filter } from 'rxjs/operators';
import { takeUntil } from 'rxjs/operators';

declare const lintelVSCodeAPI;

/**
 * Overrides is complicated, so separate it from general.ts
 */

@Component({
  changeDetection: ChangeDetectionStrategy.Default,
  providers: [DestroyService],
  selector: 'lintel-overrides',
  templateUrl: 'overrides.html',
  styleUrls: ['overrides.scss']
})

export class OverridesComponent implements OnInit {

  @ViewChild(ContextMenuComponent, { static: true }) contextMenu: ContextMenuComponent;

  overridesForm: FormGroup;

  private underConstruction: boolean;

  /** ctor */
  constructor(private actions$: Actions,
              public configs: ConfigsState,
              private contextMenuService: ContextMenuService,
              private destroy$: DestroyService,
              public extensions: ExtensionsState,
              private formBuilder: FormBuilder,
              public lintel: LintelState,
              public schema: SchemaState,
              public selection: SelectionState) {
    this.overridesForm = this.formBuilder.group({
      files: new FormArray([])
    });
    // rebuild form on selection changes
    this.handleActions$();
  }

  /** Execute context menu command */
  execute(ix: number, command: string): void {
    switch (command) {

      case 'add':
        this.configs.addOverride();
        break;

      case 'delete':
        if (!this.configs.isOverrideEmpty(ix))
          lintelVSCodeAPI.postMessage({ command: 'deleteOverride', override: ix, text: 'This override contains rules and other settings. Are you sure you want to delete it?' });
        else this.configs.deleteOverride({ ix });
        break;

    }
  }

  /** When we're ready */
  ngOnInit(): void {
    this.overridesForm.valueChanges
      .pipe(
        filter(_ => !this.underConstruction),
        takeUntil(this.destroy$)
      ).subscribe(changes => this.configs.changeOverrideFiles(changes));
    this.rebuildControls();
  }

  /** Show the context menu manually (on left click) */
  showContextMenu(event: MouseEvent, ix: number): void {
    // @see https://www.npmjs.com/package/ngx-contextmenu
    this.contextMenuService.show.next({ event, item: ix });
    event.preventDefault();
    event.stopPropagation();
  }

  // private methods

  private handleActions$(): void {
    this.actions$
      .pipe(
        filter(({ action, status }) => action.override && (status === 'SUCCESSFUL')),
        takeUntil(this.destroy$)
      )
      // NOTE: deferring the rebuild until the action is complete (add or delete)
      // is necessary becausse delete may be asynchonous if a confirm is required
      .subscribe(_ => this.rebuildControls());
  }

  private rebuildControls(): void {
    this.underConstruction = true;
    const values = this.configs.configuration.overrides?.map(override => override.files) ?? [];
    const files = this.overridesForm.controls.files as FormArray;
    while (files.length > values.length)
      files.removeAt(files.length - 1);
    while (files.length < values.length)
      files.push(new FormControl(null));
    // TODO: we don't know why { emitEvent: false } doesn't work
    files.patchValue([...values], { emitEvent: false });
    this.underConstruction = false;
  }

}
