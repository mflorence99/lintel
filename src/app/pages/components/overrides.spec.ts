import { OverridesComponent } from './overrides';

import { prepare } from './component.spec';

import { TestBed } from '@angular/core/testing';

import { async } from '@angular/core/testing';

describe('OverridesComponent', () => {

  beforeEach(async(() => prepare()));

  test('Component is created', () => {
    const fixture = TestBed.createComponent(OverridesComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  test('Override can be added', () => {
    const fixture = TestBed.createComponent(OverridesComponent);
    const component = fixture.componentInstance;
    component.selection.select({ fileName: '/home/mflorence99/lintel/package.json' });
    expect(component.configs.configuration.overrides.length).toBe(4);
    component.execute(null, 'add');
    expect(component.configs.configuration.overrides.length).toBe(5);
  });

  test('Override can be deleted', () => {
    const fixture = TestBed.createComponent(OverridesComponent);
    const component = fixture.componentInstance;
    component.selection.select({ fileName: '/home/mflorence99/lintel/package.json' });
    expect(component.configs.configuration.overrides.length).toBe(4);
    component.execute(3, 'delete');
    expect(component.configs.configuration.overrides.length).toBe(3);
  });

  test('ngOnInit', () => {
    const fixture = TestBed.createComponent(OverridesComponent);
    const component = fixture.componentInstance;
    component.ngOnInit();
    component.overridesForm.patchValue({ files: [['*.md', '*.markdown']] });
    expect(component.configs.configuration.overrides[0].files).toEqual(['*.md', '*.markdown']);
  });

});
