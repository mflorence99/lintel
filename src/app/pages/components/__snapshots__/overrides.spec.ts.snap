// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverridesComponent snapshot 1`] = `
<lintel-overrides
  __ngContext__={[Function LRootView]}
  actions$={[Function Actions]}
  cdf={[Function ViewRef]}
  configs={[Function ConfigsState]}
  contextMenu={[Function ContextMenuComponent]}
  contextMenuService={[Function ContextMenuService]}
  destroy$={[Function DestroyService]}
  extensions={[Function ExtensionsState]}
  formBuilder={[Function FormBuilder]}
  lintel={[Function LintelState]}
  overridesForm={[Function FormGroup]}
  schema={[Function SchemaState]}
  selection={[Function SelectionState]}
  underConstruction="false"
  utils={[Function Utils]}
>
  <article
    class="overrides ng-untouched ng-pristine ng-valid"
  >
    <lintel-input-array
      class="ng-untouched ng-pristine ng-valid"
    >
      <p
        class="inputs"
        style="grid-template-columns: repeat(auto-fit, 10rem);"
      >
        
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          <div
            prefix=""
          >
            <div
              prefix=""
            >
              1
            </div>
          </div>
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          
          
        </lintel-input>
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          <a
            suffix=""
          >
             More 
          </a>
          
          <div
            suffix=""
          >
            <div
              suffix=""
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cog fa-w-16"
                  data-icon="cog"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </div>
          </div>
          
        </lintel-input>
        
      </p>
      
    </lintel-input-array>
    <lintel-input-array
      class="ng-untouched ng-pristine ng-valid"
    >
      <p
        class="inputs"
        style="grid-template-columns: repeat(auto-fit, 10rem);"
      >
        
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          <div
            prefix=""
          >
            <div
              prefix=""
            >
              2
            </div>
          </div>
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          
          
        </lintel-input>
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          <a
            suffix=""
          >
             More 
          </a>
          
          <div
            suffix=""
          >
            <div
              suffix=""
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cog fa-w-16"
                  data-icon="cog"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </div>
          </div>
          
        </lintel-input>
        
      </p>
      
    </lintel-input-array>
    <lintel-input-array
      class="ng-untouched ng-pristine ng-valid"
    >
      <p
        class="inputs"
        style="grid-template-columns: repeat(auto-fit, 10rem);"
      >
        
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          <div
            prefix=""
          >
            <div
              prefix=""
            >
              3
            </div>
          </div>
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          
          
        </lintel-input>
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          
          
        </lintel-input>
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          <a
            suffix=""
          >
             More 
          </a>
          
          <div
            suffix=""
          >
            <div
              suffix=""
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cog fa-w-16"
                  data-icon="cog"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </div>
          </div>
          
        </lintel-input>
        
      </p>
      
    </lintel-input-array>
    <lintel-input-array
      class="ng-untouched ng-pristine ng-valid"
    >
      <p
        class="inputs"
        style="grid-template-columns: repeat(auto-fit, 10rem);"
      >
        
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          <div
            prefix=""
          >
            <div
              prefix=""
            >
              4
            </div>
          </div>
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          
          
        </lintel-input>
        <lintel-input
          class="ng-untouched ng-pristine ng-valid"
        >
          
          <input
            autocomplete="off"
            max="undefined"
            min="undefined"
            placeholder=""
            step="undefined"
            tabindex="0"
            type="text"
          />
          <fa-icon
            class="ng-fa-icon"
            suffix=""
            title="Remove this item"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-xxx"
              class="svg-inline--fa fa-times fa-w-11"
              data-icon="times"
              data-prefix="fas"
              role="img"
              viewBox="0 0 352 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-xxx"
              >
                Remove this item
              </title>
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                fill="currentColor"
              />
            </svg>
          </fa-icon>
          
          <a
            suffix=""
          >
             More 
          </a>
          
          <div
            suffix=""
          >
            <div
              suffix=""
            >
              <fa-icon
                class="ng-fa-icon"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cog fa-w-16"
                  data-icon="cog"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                    fill="currentColor"
                  />
                </svg>
              </fa-icon>
            </div>
          </div>
          
        </lintel-input>
        
      </p>
      
    </lintel-input-array>
    
  </article><context-menu />
</lintel-overrides>
`;
